codegenerator/cli/npm/envio/src/Debouncer.res

```diff
@@ -0,0 +1,62 @@
+type t = {
+  mutable lastRunTimeMillis: float,
+  mutable isRunning: bool,
+  mutable isAwaitingInterval: bool,
+  mutable scheduled: option<unit => promise<unit>>,
+  intervalMillis: float,
+  logger: Pino.t,
+}
+
+let make = (~intervalMillis: int, ~logger) => {
+  lastRunTimeMillis: 0.,
+  isRunning: false,
+  isAwaitingInterval: false,
+  scheduled: None,
+  intervalMillis: intervalMillis->Belt.Int.toFloat,
+  logger,
+}
+
+%%private(
+  let rec startInternal = async (debouncer: t) => {
+    switch debouncer {
+    | {scheduled: Some(fn), isRunning: false, isAwaitingInterval: false} =>
+      let timeSinceLastRun = Js.Date.now() -. debouncer.lastRunTimeMillis
+
+      //Only execute if we are passed the interval
+      if timeSinceLastRun >= debouncer.intervalMillis {
+        debouncer.isRunning = true
+        debouncer.scheduled = None
+        debouncer.lastRunTimeMillis = Js.Date.now()
```

Comment:
The way I see it, it's not a cooldown period so much as an interval, it can work either way it just makes sense in my head to have a time interval from when the last function was called, not necessarily finished.

Say a cleanup function gets called on the postgres and takes really long awaiting a transaction. By the time it's finished we can immediately execute the next scheduled function if it's passed the interval.